### PR TITLE
[NETWORK UPTIME]  Approach one

### PR DESCRIPTION
**Pull request by NoahNsimbe**
**_WHAT DOES THIS PR DO?_**
Introduces a endpoint to compute uptime using user defined configuration.

**_WHAT ISSUES ARE RELATED TO THIS PR?_**

- Jira cards
    - [AN-268-implement-approach-one](https://airqoteam.atlassian.net/browse/AN-268)
- GitHub issues
    - Closes #<enter GitHub issue number here>

**_HOW DO I TEST OUT THIS PR?_**
[Linkt to postman](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-5b3459b6-aa27-4f14-86f5-0b1ddfc45493)

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 

**_IS THERE ANY JENKINS CONSOLE LINK FOR CODE COVERAGE AND BUILD INFO?_**

**_ARE THERE ANY RELATED PRs?_**

### CODE REVIEW

**Comment by Baalmart** [comment]
Thanks @NoahNsimbe for this. 

I just have a quick feedback comment that is about strategy for managing different version of uptime calculations.
Is there a way for us to keep current version's code for the future? More like do versioning for uptime calculations?

**Context:**
In case there is a change of mind in the future (as I highly expect), we shall just switch to any version without writing any new code.

**Alternatives to shed more light on the suggestion:**

1. URL versioning: In this approach, the version number is included in the URL for the API endpoint. For example, the endpoint for version 1 of the API might be /api/v1/data and the endpoint for version 2 might be /api/v2/data.
2. Header versioning: In this approach, the version number is included in the request headers. For example, the client might send a request with a header Accept-Version: v2 to indicate that it wants to use version 2 of the API.
3. Query string versioning: In this approach, the version number is included in the query string of the URL. For example, the endpoint for version 1 of the API might be /api/data and the endpoint for version 2 might be /api/data?version=2.
4. etc....


cc: @NsubugaAnold @sserurich

**Comment by NoahNsimbe** [comment]
> Thanks @NoahNsimbe for this.
> 
> I just have a quick feedback comment that is about strategy for managing different version of uptime calculations. Is there a way for us to keep current version's code for the future? More like do versioning for uptime calculations?
> 
> **Context:** In case there is a change of mind in the future (as I highly expect), we shall just switch to any version without writing any new code.
> 
> **Alternatives to shed more light on the suggestion:**
> 
> 1. URL versioning: In this approach, the version number is included in the URL for the API endpoint. For example, the endpoint for version 1 of the API might be /api/v1/data and the endpoint for version 2 might be /api/v2/data.
> 2. Header versioning: In this approach, the version number is included in the request headers. For example, the client might send a request with a header Accept-Version: v2 to indicate that it wants to use version 2 of the API.
> 3. Query string versioning: In this approach, the version number is included in the query string of the URL. For example, the endpoint for version 1 of the API might be /api/data and the endpoint for version 2 might be /api/data?version=2.
> 4. etc....
> 
> cc: @NsubugaAnold @sserurich

Hi @Baalmart, this is still work in progress.

**Comment by NoahNsimbe** [comment]
> > Thanks @NoahNsimbe for this.
> > I just have a quick feedback comment that is about strategy for managing different version of uptime calculations. Is there a way for us to keep current version's code for the future? More like do versioning for uptime calculations?
> > **Context:** In case there is a change of mind in the future (as I highly expect), we shall just switch to any version without writing any new code.
> > **Alternatives to shed more light on the suggestion:**
> > 
> > 1. URL versioning: In this approach, the version number is included in the URL for the API endpoint. For example, the endpoint for version 1 of the API might be /api/v1/data and the endpoint for version 2 might be /api/v2/data.
> > 2. Header versioning: In this approach, the version number is included in the request headers. For example, the client might send a request with a header Accept-Version: v2 to indicate that it wants to use version 2 of the API.
> > 3. Query string versioning: In this approach, the version number is included in the query string of the URL. For example, the endpoint for version 1 of the API might be /api/data and the endpoint for version 2 might be /api/data?version=2.
> > 4. etc....
> > 
> > cc: @NsubugaAnold @sserurich
> 
> Hi @Baalmart, this is still work in progress. I shall notify you once its ready for review

File: src/device-monitoring/helpers/uptime.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        series = pd.Series(dtype=float)
+
+        series["uptime"] = 100
+        series["downtime"] = 0
+
+        if data_points < self.__data_points_threshold:
+            series["uptime"] = (data_points / self.__data_points_threshold) * 100
+            series["downtime"] = 100 - series["uptime"]
+
+        return series
```
Comment (NoahNsimbe) [inline src/device-monitoring/helpers/uptime.py:232]: Computing uptime using data points threshold

File: src/device-monitoring/helpers/request_validators.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+validate_request_params = request_validator_factory(
+    input_source=lambda: request.args,
+    info="query parameters",
+)
+
+validate_request_json = request_validator_factory(
+    input_source=lambda: request.get_json(),
+    info="json body",
+)
```
Comment (NoahNsimbe) [inline src/device-monitoring/helpers/request_validators.py:212]: request validators

File: src/device-monitoring/routes.py [hunk 3]
```diff
@@ -10,4 +10,6 @@
 
 DEVICE_UPTIME_LEADERBOARD = f"{BASE_URL}/device/uptime/leaderboard"
 
+UPTIME_BASE_URL = f"/api/v2/monitor/uptime"
+
```
Comment (NoahNsimbe) [inline src/device-monitoring/routes.py:14]: v2 endpoint to return uptime

File: src/device-monitoring/app.py [hunk 4]
```diff
@@ -101,6 +104,12 @@ def handle_exception(error):
     return jsonify({"message": "Error occurred. Contact support"}), 500
 
 
+@app.errorhandler(NotFound)
+def handle_404_exception(error):
+    print(error)
+    return jsonify({"message": "Requested Url not found on this service"}), 404
```
Comment (NoahNsimbe) [inline src/device-monitoring/app.py:110]: global 404 error handling.

**Comment by NoahNsimbe** [comment]
Hi @Baalmart 
This PR is ready for review.

**Pull request merged**